src/Api/Tools/Controllers/ReportsController.cs

```diff
@@ -0,0 +1,150 @@
+                        CollectionName = tCollect.Item1.Name,
+                        GroupId = x.Id,
+                        GroupName = groupNameDictionary[x.Id],
+                        ReadOnly = x.ReadOnly,
+                        HidePasswords = x.HidePasswords,
+                        Manage = x.Manage,
+                        CipherIds = collectionItems.ToList(),
+                        ItemCount = collectionItems.Count()
+                    });
+                accessDetails.AddRange(groupDetails);
+            }
+
+            // All collections assigned to users and their permissions
+            if (tCollect.Item2.Users.Count() > 0)
+            {
+                var userCollectionDetails = tCollect.Item2.Users.Select(x =>
+                    new MemberAccessReportAccessDetails
+                    {
+                        CollectionId = tCollect.Item1.Id,
+                        CollectionName = tCollect.Item1.Name,
+                        UserGuid = x.Id,
+                        ReadOnly = x.ReadOnly,
+                        HidePasswords = x.HidePasswords,
+                        Manage = x.Manage,
+                        CipherIds = collectionItems.ToList(),
+                        ItemCount = collectionItems.Count()
+                    });
+                accessDetails.AddRange(userCollectionDetails);
+            }
+        }
+
+        // Loop through the org users and populate report and access data
+        var orgAbility = await _applicationCacheService.GetOrganizationAbilityAsync(orgId);
+        var memberAccessReport = new List<MemberAccessReportModel>();
+        foreach (var user in orgUsers)
+        {
+            var report = new MemberAccessReportModel
+            {
+                UserName = user.OrganizationUserUserDetails.Name,
+                Email = user.OrganizationUserUserDetails.Email,
+                TwoFactorEnabled = user.TwoFactorEnabled,
+                // Both the user's ResetPasswordKey must be set and the organization can UseResetPassword
+                // var orgAbility = await _applicationCacheService.GetOrganizationAbilityAsync(orgId); 
+                AccountRecoveryEnabled = !string.IsNullOrEmpty(user.OrganizationUserUserDetails.ResetPasswordKey) && orgAbility.UseResetPassword
+            };
+
+            var userAccessDetails = new List<MemberAccessReportAccessDetails>();
+            if (user.OrganizationUserUserDetails.Groups.Any())
+            {
+                var userGroups = accessDetails.Where(x => user.OrganizationUserUserDetails.Groups.Contains(x.GroupId.GetValueOrDefault()));
+                userAccessDetails.AddRange(userGroups);
+            }
+
+            if (user.OrganizationUserUserDetails.Collections.Any())
+            {
+                var userCollections = accessDetails.Where(x => user.OrganizationUserUserDetails.Collections.Any(y => x.CollectionId == y.Id && x.UserGuid == user.OrganizationUserUserDetails.Id));
```

Comment:
This is a little hard to follow, something that might simplify it is to remove the `collectionId` comparison. I think that if they have an `accessDetails` entry with their userId, we can be sure that they have access to that collection (otherwise it wouldn't have been created in the first place).